### ISSUE TITLE
merge dictionaries recursively

### ISSUE THREAD

**Issue by AndreaCrotti** [pytoolz/toolz#281]
Sometimes it would be nice to be able to merge dictionaries recursively, for example this:

In [19]: toolz.merge({1: {2: 3}}, {1: {3: 4}})

which normally results in:

Out[19]: {1: {3: 4}}

could instead be {1: {2: 3, 3: 4}}

This is useful for example if I want to have configuration files (in YAML maybe) that override each other without having to redefine all the values.

I think it would be quite easy to do (maybe an argument recursive=False by default), what do you think?

**Comment by mrocklin** [comment]
Consider `merge_with`

```python
In [3]: merge_with(merge, {1: {2: 3}}, {1: {3: 4}})
Out[3]: {1: {2: 3, 3: 4}}
```

**Comment by AndreaCrotti** [comment]
Ah nice I actually tried that but didn't try it well obviously..

The only thing is that it only works on the first level anyway it doesn't go recursively deep.
Another option could be to have a "levels=int" into merge default to 0 that tells how many levels to go to do the merge..

**Comment by mrocklin** [comment]
One could always do the following for two-level merging  :)

```python
merge_with(merge_with(merge), dicts))
```

I'm a little wary of adding kwargs for all of the various cases because it might explode out the API.  Perhaps this case is common enough to warrant it?  I don't know.

**Comment by AndreaCrotti** [comment]
Yes wlel @mrocklin but for this particular usecase I would not really know how many levels there can be, I just need a recursive solution..

in the past I actually already implemented that somewhere and it's not that hard, but yeah we have to see if it's not just me needing something like this..

**Comment by mrocklin** [comment]
You could also implement this and we could put it in the sandbox module so
that others could play with it and, if it sees a lot of use we would
promote it to the normal namespace.

**Comment by mrucci** [comment]
That should do it, although obviously more efficient solutions definitely exist.

```python
def deep_merge(a, b):
    """ Merge two dictionaries recursively. """
    def merge_values(k, v1, v2):                                                
        if isinstance(v1, dict) and isinstance(v2, dict):
            return k, deep_merge(v1, v2)
        else:
            return k, v2
    a_keys = set(a.keys())
    b_keys = set(b.keys())
    pairs = [merge_values(k, a[k], b[k]) for k in a_keys & b_keys] \
        + [(k, a[k]) for k in a_keys - b_keys] \
        + [(k, b[k]) for k in b_keys - a_keys]
    return dict(pairs)
```

PS: toolz is coolz

**Comment by robsonpeixoto** [comment]
Thanks @mrucci

**Comment by istrupin** [comment]
Any more thoughts of getting this into the sandbox module @mrocklin?  I think this would be pretty helpful.
